jbi/queue.py

```diff
@@ -0,0 +1,315 @@
+"""This `queue` module stores Bugzilla webhook messages that we failed to sync
+to Jira.
+
+As Bugzilla sends us webhook messages, we want to eagerly accept them and
+return a `200` response so that we don't prevent it from sending new messages.
+But if we fail to sync a bug, we want to keep the message so we can retry it
+later. We also want to store any messages that might be successfuly synced, but
+were preceded by a message that wasn't synced.
+
+Classes:
+    - QueueItem: An entry in the dead letter queue, containing information
+      about the payload, timestamp, and any associated errors when attempting
+      to sync the bug.
+    - PythonException: Information about any exception that occured when
+      syncing a bug, stored along with the item.
+    - DeadLetterQueue: Class representing the dead letter queue system, providing methods
+      for adding, retrieving, and managing queue items. Supports pluggable backends.
+    - QueueBackend: Abstract base class defining the interface for a DeadLetterQueue backend.
+    - MemoryBackend: Implementation of a QueueBackend that stores messages in memory.
+    - FileBackend: Implementation of a QueueBackend that stores messages in files.
+    - InvalidQueueDSNError: Exception raised when an invalid queue DSN is provided.
+
+"""
+
+import bisect
+import itertools
+import logging
+import shutil
+import tempfile
+import traceback
+from abc import ABC, abstractmethod
+from collections import defaultdict
+from datetime import datetime, timezone
+from functools import lru_cache
+from pathlib import Path
+from typing import Optional
+from urllib.parse import ParseResult, urlparse
+
+from dockerflow import checks
+from pydantic import BaseModel, Field
+
+from jbi import bugzilla
+from jbi.environment import DLQUrl, get_settings
+
+logger = logging.getLogger(__name__)
+
+
+class PythonException(BaseModel, frozen=True):
+    type: str
+    description: str
+    details: str
+
+    @classmethod
+    def from_exc(cls, exc: Exception):
+        return PythonException(
+            type=exc.__class__.__name__,
+            description=str(exc),
```

Comment:
As said in #903, since `HTTPError` are very opaque, for easier debugging I think we could dump the response status (and body?) for `HTTPError`, with a `isinstance(exc, HTTPError) and hasattr(exc, "response")`

Reply:
To be covered in a followup, based on your approval message